src/error.rs

```diff
@@ -1,34 +1,40 @@
 //! Represents errors that can happen during a method call
 
 use failure_derive::Fail;
-use serde_derive::Deserialize;
+use serde_derive::{Deserialize, Serialize};
+use serde_json::Value;
+use std::collections::HashMap;
 
 /// Convenience type for defining `Result`s
 pub type Result<T> = std::result::Result<T, Error>;
 
 /// An error returned by the API
-#[derive(Fail, Deserialize, Debug, Eq, PartialEq, Clone)]
+#[derive(Fail, Deserialize, Serialize, Debug, PartialEq, Clone)]
 #[fail(display = "API Error #{}: {}", error_code, error_msg)]
 pub struct APIError {
     error_code: u64,
     error_msg: String,
+    #[serde(flatten)]
+    extra: HashMap<String, Value>,
 }
 
 impl APIError {
     /// Creates a new `APIError`
-    pub fn new(code: u64, msg: String) -> Self {
+    pub fn new(code: u64, msg: String, extra: HashMap<String, Value>) -> Self {
         Self {
             error_code: code,
             error_msg: msg,
+            extra: extra,
```

Comment:
You can use the field init shorthand here:
```suggestion
            extra,
```